File: day41/C++/queue.cpp
/**
 * @author: aaditkamat
 * @date: 17/02/2019
 */

/*
 * This is a linked list implementation of queues.
 */
#include "queue.h"

template<class T>
Node<T>::Node(T& element) {
  this -> value = element;
}

template<class T>
void Queue<T>::enqueue(T element) {
  Node<T>* element_node = new Node<T>(element);
  if (this -> front_node == NULL) {
    this -> front_node = element_node;
    this -> back_node = element_node;
  }
  else {
    this -> back_node -> next = element_node;
    this -> back_node = element_node;
  }
}

template<class T>
void Queue<T>::dequeue() {
  if (this -> front_node == NULL) {
    std::cerr << "Cannot remove an element from an empty queue\n";
    return;
  }
  this -> front_node = this -> front_node -> next;
}

template<class T>
T Queue<T>::front() {
  if (this -> front_node == NULL) {
    std::cerr << "Cannot return the front element of an empty queue\n";
    return NULL;
  }
  return this -> front_node -> value;
}

template<class T>
T Queue<T>::back() {
  if (this -> back_node == NULL) {
    std::cerr << "Cannot return the back element of am empty queue\n";
    return NULL;
  }
  return this -> back_node -> value;
}

File: day41/JavaScript/queue.js
/**
 * Implementation of queue usinig JS
 * @author MadhavBahl
 * @date 14/02/2019
 */

class Queue {
    constructor (size) {
        this.capacty = size;
        this.data = [];
        this.frontIndex = 0;
        this.rearIndex = 0;
    }

    front () {
        return data[this.frontIndex];
    }

    rear () {
        return data[this.rearIndex];
    }

    enqueue (element) {
        if (this.rearIndex >= this.capacty)  {
            console.log ("Overflow!");
            return -1;
        }
        this.data.unshift (element);
        console.log (element + ' added to the queue');
        this.rearIndex++;
    }

    dequeue (element) {
        if (this.rearIndex === 0) {
            console.log ("Underflow!");
            return -1;
        }
        console.log (this.data[this.rearIndex -1] + ' has been removed from the queue');
        this.rearIndex--;
        return this.data.pop ();
    }
}

const myQ = new Queue (4);

myQ.dequeue ( );
myQ.enqueue (1);
myQ.enqueue (2);
myQ.enqueue (3);
myQ.enqueue (4);
myQ.enqueue (5);
myQ.dequeue ( );